### PR TITLE
soc: arm: rpi_pico: Add support for binary info feature

### PR DESCRIPTION
**Pull request by soburi**
soc: arm: rpi_pico: Add support RaspberryPi Pico binary info feature

Binary Info embeds program meta information in flash,
which can be viewed with RaspberryPi Pico's `picotool`.

Metadata is automatically collected from pinctrl.

It can be override by the Kconfig configurations, such as

```
CONFIG_RPI_PICO_BINARY_INFO_OVERRIDE_PROGRAM_NAME="my program name"
```

When this feature is enabled, pinctrl groups are restricted to
consisting of pins with a single rpi_pico function.
In other words, SPI's MISO and MOSI can be in the same group,
but I2C's SDA cannot be included in this group.
This is rarely an issue in normal use,
and can be resolved by dividing them into separate groups.

### CODE REVIEW

**Review by mbolivar-nordic — COMMENTED** [review]
Thank you for your work!

I am not sure about the CMake variable changes here. I'd like @tejlmand to confirm it is OK.

**Comment by yonsch** [comment]
Before I comment on the changes themselves - I think this feature would be beneficial for all Zephyr users, not only for the pico. Being able to embed strings in a binary where they're accessable outside of the image is very useful. The main use case I can think of is a bootloader and an app being able to read each other's versions. Since this would require some more work I'm not sure it should block this PR for the meantime, but that is something to keep in mind.
I should also note that pico's binary_info can be used to store pin configurations, but I can't think of a use case for that.

File: soc/arm/rpi_pico/rp2/binary_info.c [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+#endif
+
+extern uint32_t __rom_region_end;
+bi_decl(bi_binary_end((intptr_t)&__rom_region_end));
+
+#ifdef PICO_PROGRAM_NAME
+bi_decl(bi_program_name((uint32_t)PICO_PROGRAM_NAME));
+#endif
+
+#ifdef PICO_BOARD
```
[the remark this answers is not in the document]
Comment (soburi) [reply soc/arm/rpi_pico/rp2/binary_info.c]: There are no strong reason to configure the settings by CMake variables. Removed it.
(It is port from Pico-SDK. It may sometimes convenient, but no strong reason to keep it.)

[the remark this answers is not in the document]
Comment (mbolivar-nordic) [reply soc/arm/rpi_pico/rp2/binary_info.c]: Thanks, resolving

**Comment by soburi** [comment]
@yonsch 
> Before I comment on the changes themselves - I think this feature would be beneficial for all Zephyr users, not only for the pico. Being able to embed strings in a binary where they're accessable outside of the image is very useful. The main use case I can think of is a bootloader and an app being able to read each other's versions. Since this would require some more work I'm not sure it should block this PR for the meantime, but that is something to keep in mind. I should also note that pico's binary_info can be used to store pin configurations, but I can't think of a use case for that.

The 'self-reference'-ish feature is interresting suggestion.
But I also think there are some difficult point. 
In the Pico's scope, the information mainly read it by BOOTSEL mode.
This is probably the same for other MCUs, and I think this function is strongly related to the "recovery mode" specification.
If there are other MCUs with similar functions, it may be possible to define an abstract interface.
If I can draw a big picture, I think it would wonderful if it could define as uf2 meta information. (Create a .uf2meta section with a linker and have the same information there. It can handle with common specification by recovery program and firmware itself!)

**Comment by yonsch** [comment]
@soburi I've created a quick demonstration of what I was thinking of: #54464

**Comment by MaureenHelm** [comment]
dev-review: depends on #54464

**Comment by soburi** [comment]
@cfriedt 

> I assume that once #54464 is merged, work on this will continue

That's right. Perhaps I will have to rethink the design accordingly.

**Comment by yonsch** [comment]
@soburi Why does this depend on #54464?
Bindesc was supposed to be a standard Zephyr replacement for binary info.
Binary info would still be useful for someone who needs to specifically use the pico format, e.g. if a bootloader expects binary info descriptors.
I think we can treat them as if they are mutually exclusive, as they serve the same purpose, and proceed with this PR regardless of bindesc.
Unless, you somehow see them coexisting somehow, which would require more planning.

**Comment by MaureenHelm** [comment]
> @cfriedt
> 
> > I assume that once #54464 is merged, work on this will continue
> 
> That's right. Perhaps I will have to rethink the design accordingly.

Has this been superseded by #54464? Can we close?

File: soc/raspberrypi/rpi_pico/common/binary_info_header.S
Comment (dsseng) [inline soc/raspberrypi/rpi_pico/common/binary_info_header.S:1]: Could this potentially be handled by a linker script like rom_start?

Comment (soburi) [reply soc/raspberrypi/rpi_pico/common/binary_info_header.S:1]: Addressed.
Thank you.

File: soc/raspberrypi/rpi_pico/common/Kconfig.binary_info [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	help
+	  Binary info is able to embed machine readable information with the binary in flash.
+	  It can read with picotool(https://github.com/raspberrypi/picotool).
+	  Binary info is generated automatically when this option is enabled,
+
+	  When this feature is enabled, pinctrl groups are restricted to consisting
+	  of pins with a single rpi_pico function.
+	  In other words, SPI's MISO and MOSI can be in the same group, but I2C's
+	  SDA cannot be included in this group. This is rarely an issue in normal use,
+	  and can be resolved by dividing them into separate groups.
```
Comment (dsseng) [inline soc/raspberrypi/rpi_pico/common/Kconfig.binary_info:16]: Does it now always pack pinctrl data into the binary? Could you please give a binary size comparison, as this seems to increase it for quite a bit

Comment (soburi) [reply soc/raspberrypi/rpi_pico/common/Kconfig.binary_info:16]: >                0x00000000100044c8                __binary_info_start = .
>  *(SORT_BY_ALIGNMENT(.binary_info.keep.*))
>  .binary_info.keep.__bi_ptr51
>                 0x00000000100044c8        0x4 zephyr/libzephyr.a(binary_info.c.obj)
>  .binary_info.keep.__bi_ptr43
>                 0x00000000100044cc        0x4 zephyr/libzephyr.a(binary_info.c.obj)
>  .binary_info.keep.__bi_ptr38
>                 0x00000000100044d0        0x4 zephyr/libzephyr.a(binary_info.c.obj)
>  .binary_info.keep.__bi_ptr33
>                 0x00000000100044d4        0x4 zephyr/libzephyr.a(binary_info.c.obj)
>  .binary_info.keep.__bi_ptr29
>                 0x00000000100044d8        0x4 zephyr/libzephyr.a(binary_info.c.obj)
>  .binary_info.keep.__bi_ptr25
>                 0x00000000100044dc        0x4 zephyr/libzephyr.a(binary_info.c.obj)
>  .binary_info.keep.__bi_ptr21
>                 0x00000000100044e0        0x4 zephyr/libzephyr.a(binary_info.c.obj)
>  .binary_info.keep.__bi_ptr17
>                 0x00000000100044e4        0x4 zephyr/libzephyr.a(binary_info.c.obj)
>  .binary_info.keep.__bi_ptr13
>                 0x00000000100044e8        0x4 zephyr/libzephyr.a(binary_info.c.obj)
>  .binary_info.keep.__bi_ptr10
>                 0x00000000100044ec        0x4 zephyr/libzephyr.a(binary_info.c.obj)
>  .binary_info.keep.__bi_ptr190
>                 0x00000000100044f0        0x4 zephyr/libzephyr.a(binary_info_pins.c.obj)
>  .binary_info.keep.__bi_ptr156
>                 0x00000000100044f4        0x4 zephyr/libzephyr.a(binary_info_pins.c.obj)
>  .binary_info.keep.__bi_ptr123
>                 0x00000000100044f8        0x4 zephyr/libzephyr.a(binary_info_pins.c.obj)
>  .binary_info.keep.__bi_ptr122
>                 0x00000000100044fc        0x4 zephyr/libzephyr.a(binary_info_pins.c.obj)
>  .binary_info.keep.__bi_ptr88
>                 0x0000000010004500        0x4 zephyr/libzephyr.a(binary_info_pins.c.obj)
>  .binary_info.keep.__bi_ptr54
>                 0x0000000010004504        0x4 zephyr/libzephyr.a(binary_info_pins.c.obj)
>  .binary_info.keep.__bi_ptr21
>                 0x0000000010004508        0x4 zephyr/libzephyr.a(binary_info_pins.c.obj)
>  .binary_info.keep.__bi_ptr20
>                 0x000000001000450c        0x4 zephyr/libzephyr.a(binary_info_pins.c.obj)
>  *(SORT_BY_ALIGNMENT(.binary_info.*))
>                 0x0000000010004510                __binary_info_end = .

0x0000000010004510 - 0x00000000100044c8 = 72 (bytes)

Although it's not zero, I don't think it's a large number to concern.
Of course, we can disable it as an option, but we would like to keep it enabled by default to be compatible with the rpi_pico ecosystem.

Comment (dsseng) [reply soc/raspberrypi/rpi_pico/common/Kconfig.binary_info:16]: Great, thanks for clarifying. This is clearly okay :)

File: soc/raspberrypi/rpi_pico/common/Kconfig.binary_info [hunk 3]
```diff
@@ -0,0 +1,177 @@
+# Copyright (c) 2024 TOKITA Hiroshi
```
[the remark this answers is not in the document]
Comment (soburi) [reply soc/raspberrypi/rpi_pico/common/Kconfig.binary_info:1]: The source was posted first in 2024.

File: soc/raspberrypi/rpi_pico/common/binary_info_header.ld [hunk 4]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright (c) 2023 TOKITA Hiroshi <tokita.hiroshi@gmail.com>
```
[the remark this answers is not in the document]
Comment (soburi) [reply soc/raspberrypi/rpi_pico/common/binary_info_header.ld:2]: The source was posted first in 2023.

File: soc/raspberrypi/rpi_pico/common/binary_info_pins.c [hunk 5]
```diff
@@ -0,0 +1,1109 @@
...
+BINARY_INFO_FROM_PINCFG_GROUP(DT_NODELABEL(pinctrl), 31, 23);
+BINARY_INFO_FROM_PINCFG_GROUP(DT_NODELABEL(pinctrl), 31, 24);
+BINARY_INFO_FROM_PINCFG_GROUP(DT_NODELABEL(pinctrl), 31, 25);
+BINARY_INFO_FROM_PINCFG_GROUP(DT_NODELABEL(pinctrl), 31, 26);
+BINARY_INFO_FROM_PINCFG_GROUP(DT_NODELABEL(pinctrl), 31, 27);
+BINARY_INFO_FROM_PINCFG_GROUP(DT_NODELABEL(pinctrl), 31, 28);
+BINARY_INFO_FROM_PINCFG_GROUP(DT_NODELABEL(pinctrl), 31, 29);
+BINARY_INFO_FROM_PINCFG_GROUP(DT_NODELABEL(pinctrl), 31, 30);
+BINARY_INFO_FROM_PINCFG_GROUP(DT_NODELABEL(pinctrl), 31, 31);
+#endif
```
[the remark this answers is not in the document]
Comment (soburi) [reply soc/raspberrypi/rpi_pico/common/binary_info_pins.c:1106]: We can't use anything other than LINE, and I don't have the habit of writing code snippets.
There's no other way.

File: soc/raspberrypi/rpi_pico/common/Kconfig.binary_info [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+config RPI_PICO_BINARY_INFO
+	bool "Generate RaspberryPi Pico binary info"
```
Comment (ajf58) [inline soc/raspberrypi/rpi_pico/common/Kconfig.binary_info]:
```suggestion
	bool "Generate Raspberry Pi Pico binary info"
```

File: soc/raspberrypi/rpi_pico/common/Kconfig.binary_info [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+	default y
+	help
+	  Binary info is able to embed machine readable information with the binary in flash.
+	  It can be read with picotool(https://github.com/raspberrypi/picotool).
```
Comment (ajf58) [inline soc/raspberrypi/rpi_pico/common/Kconfig.binary_info]:
```suggestion
	  It can be read with picotool (https://github.com/raspberrypi/picotool).
```

**Review by ajf58 — DISMISSED** [review]

File: soc/raspberrypi/rpi_pico/common/Kconfig.binary_info (hunk 6 above)
Comment (soburi) [reply soc/raspberrypi/rpi_pico/common/Kconfig.binary_info]: Fixed. Thank you for your review.

File: soc/raspberrypi/rpi_pico/common/Kconfig.binary_info (hunk 7 above)
Comment (soburi) [reply soc/raspberrypi/rpi_pico/common/Kconfig.binary_info]: ditto.

**Review by ajf58 — APPROVED** [review]

File: soc/raspberrypi/rpi_pico/common/CMakeLists.txt
Comment (dsseng) [inline soc/raspberrypi/rpi_pico/common/CMakeLists.txt:18]: `0x0binary_info_header` looks weird to me, but I'm not really into the linker magic

**Review by dsseng — APPROVED** [review]
It's alright, not that busy :). Looks great from my perspective!

**Review by fabiobaltieri — APPROVED** [review]
